http/client/src/main/java/org/wildfly/security/http/client/ElytronHttpClient.java

```diff
@@ -0,0 +1,139 @@
+ * limitations under the License.
+ */
+package org.wildfly.security.http.client;
+
+import org.wildfly.security.http.client.exception.ElytronHttpClientException;
+import org.wildfly.security.http.client.mechanism.basic.ElytronHttpClientBasicAuthMechanism;
+import org.wildfly.security.http.client.mechanism.bearer.ElytronHttpClientBearerAuthMechanism;
+import org.wildfly.security.http.client.mechanism.digest.ElytronHttpClientDigestAuthMechanism;
+import org.wildfly.security.http.client.utils.ElytronHttpClientConstants;
+import org.wildfly.security.http.client.utils.ElytronHttpClientCredentialUtils;
+import org.wildfly.security.http.client.utils.ElytronHttpClientRequestBuilder;
+import org.wildfly.security.http.client.utils.ElytronMessages;
+
+import javax.net.ssl.SSLContext;
+
+import static org.wildfly.security.http.HttpConstants.OK;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Elytron client for HTTP authentication
+ *
+ * @author <a href="mailto:kekumar@redhat.com">Keshav Kumar</a>
+ */
+public class ElytronHttpClient {
+
+    private HttpClient httpClient;
+
+    public ElytronHttpClient() {
+        this.httpClient = HttpClient.newHttpClient();
+    }
+
+    private HttpResponse getResponse(HttpRequest request) throws IOException, InterruptedException {
+        return httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+    }
+
+    private void addSSLContextToHttpClient(URI uri) throws MalformedURLException {
+        String protocol = new URL(uri.toString()).getProtocol();
+
+        if (protocol.equalsIgnoreCase("https")) {
+            SSLContext sslContext = ElytronHttpClientCredentialUtils.getSSLContext(uri);
+            httpClient = HttpClient.newBuilder().sslContext(sslContext).build();
+        }
+    }
+
+    /**
+     * Used to connect to the secured uri and return the response based on that.
+     */
+    public HttpResponse connect(String uri, String method, String body, Map<String, String> headers) throws IOException, InterruptedException, URISyntaxException {
+        URI uriPath = new URI(uri);
+        addSSLContextToHttpClient(uriPath);
+
+        HttpRequest request = ElytronHttpClientRequestBuilder.buildRequest(uriPath, method, body, headers);
+        HttpResponse response = getResponse(request);
+
+        if (response.statusCode() == OK) {
+            return response;
+        }
+
+        String authHeader = getAuthHeader(response);
+
+        if (authHeader == null) {
+            throw new ElytronHttpClientException(ElytronMessages.log.responseHeaderExtractionFailed());
+        }
+        String[] authChallenges = authHeader.split(",");
+        HttpRequest authRequest = null;
+
+        String challenge = authChallenges[0];
```

Comment:
Just a minor, I would introduce a new method here, for example String challenge = getChallenge(headers) or getChallenge(response) so we can hide the array splitting and parsing implementation details from this main method